### PR TITLE
Closes #1550 Initial Intelligence Bank integration.

### LINKED ISSUE #1550: Add IntelligenceBank integration to Text on Media
## Motivation
The University is rolling out a Digital Asset Manager called IntelligenceBank to make it easier for everyone at the university to share digital assets. Integrating this into Quickstart will make it accessible to hundreds of marketers and communicators across campus to help ensure successful adoption. 

This will also solve one of the biggest pain points for site owners: how to find good images to tell their stories.

## Proposed Resolution
The first step in adding this integration, is to make these assets available for Text on Media. As the element that has the most impact with images, it is a great place to start.

### Describe the solution you'd like
In the media field of the Text on Media, there should be two buttons as shown here:
Screen Shot 2022-05-11 at 2 43 06 PM

Clicking into the Add Arizona Digital Assets button would pull up the full IntelligenceBank Asset Browser (for authenticated users) using the Entity Browser widget (configured at /admin/structure/paragraphs_type/az_text_media/form-display). 
Screen Shot 2022-05-11 at 5 50 37 PM

If an image is selected from Arizona Digital Assets, this is how it would display on the form:
Screen Shot 2022-05-11 at 2 46 16 PM

When selecting an image, there will be two options (Download and CDN), supporting each of these can be separate tickets.
After selecting an image, there will be a Settings screen that should _only_ have two options: Alternative Text and Title.

If an image is selected from Media, this is how it would display on the form:
Screen Shot 2022-05-11 at 2 47 08 PM

In either case, the alternative option is not displayed unless the image is removed.
The Text on Media should now display the same regardless of the image source.
Support for the "Media Aspect Ratio" space around content setting will require further discussion and can likely be a follow-up.

### Describe alternatives you've considered
The ideal experience would be to have the IntelligenceBank assets directly selectable using the core Media library, but we don't know the timeline for this to get implemented by their developers. (https://www.drupal.org/project/intelligencebank/issues/3177500) The current form of Media Library integration only displays assets that have already been imported through the Entity Browser. This alternative is our next best implementation.

### Roles and Permissions considerations
Same permissions as Media

### LINKED ISSUE #1551: IntelligenceBank integration
**Do not disclose security vulnerabilities here, please email us instead: 
az-digital-security@list.arizona.edu**

## Motivation
Why the feature request was filed, example use cases, etc.

### Roles and Permissions considerations
A clear and concise description of how each of the following roles would be impacted by this change:
- Anonymous user
- Authenticated user
- Content editor
- Content administrator
- Administrator

### PR DESCRIPTION
**Pull request by trackleft**
## Description
Adds the ability to download images from the DAM and manage as either remote media or local media.

## Related issues
Closes #1550 

Spike: https://github.com/az-digital/az_quickstart/issues/1313
EPIC: https://github.com/az-digital/az_quickstart/issues/1551
MVP Issue: https://github.com/az-digital/az_quickstart/issues/1550
MVP PR https://github.com/az-digital/az_quickstart/pull/1592
Follow-up issue: https://github.com/az-digital/az_quickstart/issues/1686
Follow-up issue: https://github.com/az-digital/az_quickstart/issues/1689
Follow-up issue: https://github.com/az-digital/az_quickstart/issues/1688

## How to test
Enable az_dam module.
Go to /admin/content/media
Click on  [+ Add DAM assets]
In the url field click the custom url checkbox
In the url field add assets.arizona.edu
click sign in 
Click sign in with net id
Close once you are signed in, go back to the Add DAM assets page and click continue, or sign in and continue
Now you should be able to find an image and Download it.

## Types of changes
New DAM integration feature
### Arizona Quickstart (install profile, custom modules, custom theme)
- Patch release changes
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [x] Critical institutional link or brand change
- Minor release changes
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- [ ] Other or unknown

### Drupal core
- Patch release changes
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major or minor level update
- [ ] Other or unknown

### Drupal contrib projects
- Patch release changes
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major level update
- [ ] Other or unknown

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: modules/custom/az_dam/config/quickstart/ib_dam.settings.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+dependencies:
+  enforced:
+    module:
+      - az_dam
+debug: true
```
Comment (trackleft) [inline modules/custom/az_dam/config/quickstart/ib_dam.settings.yml:5]: false

Comment (trackleft) [inline modules/custom/az_dam/config/quickstart/ib_dam.settings.yml:5]:
```suggestion
debug: false
```

**Comment by joshuasosa** [comment]
I tested this locally on DDEV and encountered a white screen of death when clicking to Download an image. I also tried it on the ProboCI site and got the same issue. Looking at reports gave some error info.

> Unable to save local file:
> This value should be of the correct primitive type.

> TypeError: Argument 1 passed to Drupal\ib_dam\AssetFormatter\AssetFormatterManager::create() must implement interface Drupal\ib_dam\Asset\AssetInterface, bool given, called in /src/az-quickstart-scaffolding/web/modules/contrib/intelligencebank/modules/ib_dam_media/src/Plugin/EntityBrowser/Widget/IbDamEbWidget.php on line 328 in Drupal\ib_dam\AssetFormatter\AssetFormatterManager::create() (line 29 of /src/az-quickstart-scaffolding/web/modules/contrib/intelligencebank/src/AssetFormatter/AssetFormatterManager.php)

After encountering that error, I tried clicking 'back' on my browser and it took me back to the screen to login with my credentials, which is annoying. It'd be better if it stayed connected.

Also, a related issue I'd have with the usability of this integration is that assets in the "University Photo Library" aren't organized well and generally don't provide useful titles to be able to search on. I saw some Old Main photos on the first page and tried searching for Old Main, but got no results. Having a repository of 6k images with poor searchability makes that part of the repository almost useless for users wanting to find a relevant image. A title of "Now_41" or "UA Campus (25)" gives me no context to know what I'm looking at or what I'd even search for to find that sort of image. Pagination is very slow, which makes just browsing through the images a tedious user experience.

Even after resolving errors, until the user experience is improved for actually finding relevant assets at least through better contextual titles, I'd suggest not enabling this feature request by default.

**Comment by trackleft** [comment]
I plan on at least attempting creating a patch against the intelligence bank project on drupal.org that solves the login shenanigans.

File: modules/custom/az_dam/az_dam.info.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+name: 'Quickstart DAM'
+type: module
+description:  'The connector for the centralized digital asset management system (IntelligenceBank DAM).'
```
Comment (joeparsons) [inline modules/custom/az_dam/az_dam.info.yml:3]:
```suggestion
description: 'The connector for the centralized digital asset management system (IntelligenceBank DAM).'
```

**Comment by trackleft** [comment]
Move to a follow-up for allowing cdn usage.
```
/**
 * Implements hook_form_alter().
 *
 * Add custom submit handler to form.
 * Allow using remote media.
 * See
 * @link https://git.drupalcode.org/project/intelligencebank/-/blob/8.x-2.x/modules/ib_dam_media/src/Plugin/EntityBrowser/Widget/IbDamEbWidget.php#L249 @endlink.
 */
function az_dam_form_entity_browser_az_dam_form_alter(&$form, FormStateInterface $form_state, $form_id) {
  $form['widget']['ib_dam_app_el']['#allow_embed'] = TRUE;
  $widget_context = $form_state->get(['entity_browser', 'widget_context']);
  $widget_context['ib_dam_media']['allow_embed'] = TRUE;
  $form_state->set(['entity_browser', 'widget_context'], $widget_context);
  $form['#submit'][] = 'az_dam_form_submit';
}
```

**Comment by trackleft** [comment]
Updated name of module from "Quickstart DAM" to "Quickstart Arizona Digital Asset Library" and added to "experimental modules" group

Removed remote asset hosting functionality (for now)

Instructions for enabling
Go to `/admin/modules`
Enable Quickstart Digital Asset Library
image
image

Still need to work out permissions, but that is fairly trivial.

Go to `/admin/content/media`
image

Click on "+ Add Arizona Digital Assets"
image

Select an asset
image

File: modules/custom/az_digital_asset_library/config/quickstart/ib_dam_media.settings.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    media_type: az_audio
+  video:
+    source_type: video
+    media_type: az_video
+  file:
+    source_type: file
+    media_type: az_document
+  embed:
+    source_type: embed
+    media_type: ib_dam_embed
```
[the remark this answers is not in the document]
Comment (camikazegreen) [reply modules/custom/az_digital_asset_library/config/quickstart/ib_dam_media.settings.yml:20]: I can add my approval after this change is committed.

**Comment by joeparsons** [comment]
Would be good to merge main into this one when possible.

**Review by camikazegreen — APPROVED** [review]

File: modules/custom/az_digital_asset_library/config/quickstart/ib_dam.settings.yml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+dependencies:
+  enforced:
+    module:
+      - az_digital_asset_library
+debug: false
+staging: true
```
Comment (camikazegreen) [inline modules/custom/az_digital_asset_library/config/quickstart/ib_dam.settings.yml:6]:
```suggestion
staging: false
```
According to Garrett, the staging environment should no longer be required.

**Review by tadean — APPROVED** [review]

**Pull request merged**